sa/sa.go

```diff
@@ -1328,3 +1297,117 @@ func (ssa *SQLStorageAuthority) GetOrder(ctx context.Context, req *sapb.OrderReq
+		return nil, berrors.InternalServerError("no names received")
+	}
+
+	params := make([]interface{}, len(names))
+	qmarks := make([]string, len(names))
+	for i, name := range names {
+		id := core.AcmeIdentifier{Type: core.IdentifierDNS, Value: name}
+		idJSON, err := json.Marshal(id)
+		if err != nil {
+			return nil, err
+		}
+		params[i] = string(idJSON)
+		qmarks[i] = "?"
+	}
+
+	var auths []*core.Authorization
+	_, err := ssa.dbMap.Select(
+		&auths,
+		fmt.Sprintf(`SELECT %s FROM %s
+	WHERE registrationID = ? AND
+	expires > ? AND
+	status = ? AND
+	identifier IN (%s)`, authzFields, table, strings.Join(qmarks, ",")),
+		append([]interface{}{registrationID, now, status}, params...)...)
+	if err != nil {
+		return nil, err
+	}
+
+	byName := make(map[string]*core.Authorization)
+	for _, auth := range auths {
+		// No real life authorizations should have a nil expires. If we find them,
+		// don't consider them valid.
+		if auth.Expires == nil {
+			continue
+		}
+		if auth.Identifier.Type != core.IdentifierDNS {
+			return nil, fmt.Errorf("unknown identifier type: %q on authz id %q", auth.Identifier.Type, auth.ID)
+		}
+		existing, present := byName[auth.Identifier.Value]
+		if !present || auth.Expires.After(*existing.Expires) {
+			byName[auth.Identifier.Value] = auth
+		}
+	}
+
+	return byName, nil
+}
+
+func (ssa *SQLStorageAuthority) getPendingAuthorizations(ctx context.Context, registrationID int64, names []string, now time.Time) (map[string]*core.Authorization, error) {
+	return ssa.getAuthorizations(ctx, pendingAuthorizationTable, string(core.StatusPending), registrationID, names, now)
+}
+
+func authzMapToPB(m map[string]*core.Authorization) (*sapb.Authorizations, error) {
+	resp := &sapb.Authorizations{}
+	for k, v := range m {
+		authzPB, err := bgrpc.AuthzToPB(*v)
+		if err != nil {
+			return nil, err
+		}
+		// Make a copy of k because it will be reassigned with each loop.
+		kCopy := k
+		resp.Authz = append(resp.Authz, &sapb.Authorizations_MapElement{Domain: &kCopy, Authz: authzPB})
+	}
+	return resp, nil
+}
+
+// GetAuthorizations returns a map of valid or pending authorizations for as many names as possible
+func (ssa *SQLStorageAuthority) GetAuthorizations(ctx context.Context, req *sapb.GetAuthorizationsRequest) (*sapb.Authorizations, error) {
+	authz, err := ssa.GetValidAuthorizations(ctx, *req.RegistrationID, req.Domains, time.Unix(0, *req.Now))
+	if err != nil {
+		return nil, err
+	}
+	if len(authz) == len(req.Domains) {
+		return authzMapToPB(authz)
+	}
+
+	// remove names we already have authz for
+	for i, l := 0, len(req.Domains); i < l; i++ {
+		if _, present := authz[req.Domains[i]]; present {
+			req.Domains = append(req.Domains[:i], req.Domains[i+1:]...)
+			i--
+			l--
+		}
+	}
+	pendingAuthz, err := ssa.getPendingAuthorizations(ctx, *req.RegistrationID, req.Domains, time.Unix(0, *req.Now))
+	if err != nil {
+		return nil, err
+	}
+	// merge pending into valid
+	for n, a := range pendingAuthz {
```

Comment:
Let's use a more descriptive name, like domain, since `n` makes it look like the loop variable is number-valued rather than string-valued.